extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx

```diff
@@ -1,14 +1,112 @@
+   const [estimatedDelivery, setEstimatedDelivery] = useState<string>(
+      getDefaultEstimatedDeliveryDate()
+   );
+
+   useEffect(() => {
+      const newDeliveryGroup =
+         deliveryGroupTarget?.deliveryOptions?.find(
+            (option) =>
+               option.handle ===
+               deliveryGroupTarget?.selectedDeliveryOption?.handle
+         ) ?? null;
+      if (
+         isShippingOption(newDeliveryGroup) &&
+         newDeliveryGroup?.deliveryEstimate?.timeInTransit?.upper
+      ) {
+         const newEstimatedDeliveryDate = getEstimatedDeliveryDate(
+            newDeliveryGroup.deliveryEstimate.timeInTransit.upper
+         );
+         if (newEstimatedDeliveryDate !== estimatedDelivery) {
+            setEstimatedDelivery(newEstimatedDeliveryDate ?? '');
+         }
+      }
+   }, [
+      deliveryGroupTarget?.selectedDeliveryOption?.handle,
+      deliveryGroupTarget?.deliveryOptions,
+      estimatedDelivery,
+   ]);
+
+   useUpdateAttribute({
+      key: 'Estimated Delivery Date',
+      value: estimatedDelivery,
+   });
+   return (
       <BlockStack>
          <CheckoutHint />
          <VatSection />
       </BlockStack>
-   )
-);
+   );
+}
+
+function getDefaultEstimatedDeliveryDate() {
+   // One week from now
+   const estimatedDeliveryDate = new Date(Date.now() + 7 * 24 * 60 * 60 * 1000);
+   // Format: YYYY-MM-DD
+   return estimatedDeliveryDate.toISOString().split('T')[0];
+}
+
+function getEstimatedDeliveryDate(secondsToDelivery: number): string {
+   const maxDays = Math.ceil(secondsToDelivery / 60 / 60 / 24);
+   const daysUntilDelivery = workDaysToDaysFromNow(maxDays);
+
+   const estimatedDeliveryDate = new Date(
+      Date.now() + daysUntilDelivery * 24 * 60 * 60 * 1000
+   );
+   // Format: YYYY-MM-DD
+   return estimatedDeliveryDate.toISOString().split('T')[0];
+}
+
+/**
+ * Adopted from https://stackoverflow.com/a/40739060/12671411
+ */
+function workDaysToDaysFromNow(inDaysToAdd: number) {
+   let daysToAdd = inDaysToAdd;
+   // Get the day of the week as a number (0 = Sunday, 1 = Monday, .... 6 = Saturday)
+   const dayOfWeek = new Date().getDay();
+   // If the current day is Sunday add one day
+   if (dayOfWeek === 0) {
+      daysToAdd++;
+   }
+   // If the start date plus days to add falls on or after the closest Saturday calculate weekends
+   if (dayOfWeek + daysToAdd >= 6) {
+      // Subtract days in current working week from work days
+      const remainingWorkDays = daysToAdd - (5 - dayOfWeek); // 7
+      // Add current working week's weekend
+      daysToAdd += 2; // 8
+      if (remainingWorkDays > 5) {
+         // Add two weekend days for each week remaining
+         daysToAdd += 2 * Math.floor(remainingWorkDays / 5);
+         10;
+         // Exclude final weekend if remainingWorkDays resolves to an exact number of weeks
+         if (remainingWorkDays % 5 === 0) {
+            daysToAdd -= 2;
+         }
+      }
+   }
+   return daysToAdd;
+}
```

Comment:
I would probably do something like:

```suggestion
function oneWeekFromNow() {
   return formatDate(new Date(Date.now() + 7 * 24 * 60 * 60 * 1000));
}

function estimatedDeliveryDate(deliveryEstimate: DeliveryEstimate) {
   return formatDate(new Date(
      Date.now() + deliveryEstimate.timeInTransit.upper * 1000
   ));
}

/** Format: YYYY-MM-DD */
function formatDate(date: Date) {
   return date.toISOString().split('T')[0];
}
```